src/script/interpreter.cpp

```diff
@@ -982,76 +982,44 @@ namespace {
-        txTo(txToIn), scriptCode(scriptCodeIn), nIn(nInIn),
+    CTransactionSignatureInnerSerializer(const CTransaction &txToIn, int nHashTypeIn) :
+        txTo(txToIn),
         fAnyoneCanPay(!!(nHashTypeIn & SIGHASH_ANYONECANPAY)),
         fHashSingle((nHashTypeIn & 0x1f) == SIGHASH_SINGLE),
         fHashNone((nHashTypeIn & 0x1f) == SIGHASH_NONE) {}
 
-    /** Serialize the passed scriptCode, skipping OP_CODESEPARATORs */
-    template<typename S>
-    void SerializeScriptCode(S &s, int nType, int nVersion) const {
-        CScript::const_iterator it = scriptCode.begin();
-        CScript::const_iterator itBegin = it;
-        opcodetype opcode;
-        unsigned int nCodeSeparators = 0;
-        while (scriptCode.GetOp(it, opcode)) {
-            if (opcode == OP_CODESEPARATOR)
-                nCodeSeparators++;
-        }
-        ::WriteCompactSize(s, scriptCode.size() - nCodeSeparators);
-        it = itBegin;
-        while (scriptCode.GetOp(it, opcode)) {
-            if (opcode == OP_CODESEPARATOR) {
-                s.write((char*)&itBegin[0], it-itBegin-1);
-                itBegin = it;
-            }
-        }
-        if (itBegin != scriptCode.end())
-            s.write((char*)&itBegin[0], it-itBegin);
-    }
-
     /** Serialize an input of txTo */
     template<typename S>
     void SerializeInput(S &s, unsigned int nInput, int nType, int nVersion) const {
-        // In case of SIGHASH_ANYONECANPAY, only the input being signed is serialized
-        if (fAnyoneCanPay)
-            nInput = nIn;
         // Serialize the prevout
         ::Serialize(s, txTo.vin[nInput].prevout, nType, nVersion);
         // Serialize the script
         assert(nInput != NOT_AN_INPUT);
-        if (nInput != nIn)
-            // Blank out other inputs' signatures
-            ::Serialize(s, CScript(), nType, nVersion);
-        else
-            SerializeScriptCode(s, nType, nVersion);
+        // Blank all inputs' signatures
+        // The scriptOut for the input being signed is
+        // serialized in the outer serializer
+        ::Serialize(s, CScript(), nType, nVersion);
         // Serialize the nSequence
-        if (nInput != nIn && (fHashSingle || fHashNone))
-            // let the others update at will
+        if (fHashSingle || fHashNone) {
+            // let all inputs update at will
             ::Serialize(s, (int)0, nType, nVersion);
-        else
+        } else {
             ::Serialize(s, txTo.vin[nInput].nSequence, nType, nVersion);
+        }
     }
 
     /** Serialize an output of txTo */
     template<typename S>
     void SerializeOutput(S &s, unsigned int nOutput, int nType, int nVersion) const {
-        if (fHashSingle && nOutput != nIn)
-            // Do not lock-in the txout payee at other indices as txin
-            ::Serialize(s, CTxOut(), nType, nVersion);
-        else
-            ::Serialize(s, txTo.vout[nOutput], nType, nVersion);
+        ::Serialize(s, txTo.vout[nOutput], nType, nVersion);
```

Comment:
Because I don't understand ``fHashSingle`` I don't understand this change. Since there is no ``nIn`` for inner hashes, shouldn't the if/else remain here with the if condition being just ``if (fHashSingle) { … } else { … }`` ?

Another angle on the same concern: it looks like ``fHashSingle`` mode *does* influence inner hashes, since there are branches on it inside ``SerializeInput``. So why does it not *also* influence ``SerializeOutput`` for consistency?